projects/interfaces/main.go

```diff
@@ -0,0 +1,128 @@
+package main
+
+import (
+	"io"
+	"unicode"
+)
+
+// ---------- Implementing our own bytes.Buffer
+
+// the original Buffer type definition:
+//
+//	type Buffer struct {
+//		buf      []byte // contents are the bytes buf[off : len(buf)]
+//		off      int    // read at &buf[off], write at &buf[len(buf)]
+//		lastRead readOp // last read operation, so that Unread* can work correctly.
+//	}
+type OurByteBuffer struct {
+	buf []byte
+}
+
+// the original Buffer Bytes() method signature:
+// func (b *Buffer) Bytes() []byte
+func (b *OurByteBuffer) Bytes() []byte {
+	// simply return the buffer
+	return b.buf
+}
+
+// the original Buffer Write() method signature:
+// func (b *Buffer) Write(p []byte) (n int, err error)
+func (b *OurByteBuffer) Write(p []byte) (int, error) {
+	// Q: Why is no error returned from append(?)
+	// A: The append function in Go doesn't generally return an error because it performs a safe operation that cannot fail in most cases.
+	// The reason for this is related to how Go handles memory allocation and slice operations.
+
+	// append the byte slice values (passed to the function) to the buffer
+	b.buf = append(b.buf, p...)
+
+	// return the number of bytes that were appended, and an error... although it can never be not nil(??)
```

Comment:
I don't understand this
I would have thought because we are doing an append that it COULD fail, eg we don't have enough memory to make a new copy (because isn't an append a total new copy?)

Reply:
Have a think about this again in light of my other comment about Go "pretending" allocation can't fail :)